src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/KNNEngine.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package com.amazon.opendistroforelasticsearch.knn.index.util;
+
+import com.amazon.opendistroforelasticsearch.knn.index.KNNMethod;
+import com.amazon.opendistroforelasticsearch.knn.index.KNNMethodContext;
+import com.amazon.opendistroforelasticsearch.knn.index.SpaceType;
+
+import static com.amazon.opendistroforelasticsearch.knn.common.KNNConstants.NMSLIB_NAME;
+
+
+/**
+ * KNNEngine provides the functionality to validate and transform user defined indices into information that can be
+ * passed to the respective k-NN library's JNI layer.
+ */
+public enum KNNEngine implements KNNLibrary {
+    NMSLIB(NMSLIB_NAME, Nmslib.INSTANCE);
+
+    public static final KNNEngine DEFAULT = NMSLIB;
+
+    /**
+     * Constructor for KNNEngine
+     *
+     * @param name name of engine
+     * @param knnLibrary library the engine uses
+     */
+    KNNEngine(String name, KNNLibrary knnLibrary) {
+        this.name = name;
+        this.knnLibrary = knnLibrary;
+    }
+
+    private String name;
+    private KNNLibrary knnLibrary;
+
+    /**
+     * Get the engine
+     *
+     * @param name of engine to be fetched
+     * @return KNNEngine corresponding to name
+     */
+    public static KNNEngine getEngine(String name) {
```

Comment:
on second thought, why do we need this? It is responsibility of caller to know what Enum to use. This could be helper method outside Enum. what do you think?

Reply:
Understand. Shall we move this to helper method?

Reply:
So, you are suggesting get rid of the enum, correct?

Reply:
No, thats fine. Let this be here. It is overkill to add new class and method. Just curious valueOf() will not be useful to get instance?

Reply:
I think we could do that, but then the name the user passes in would need to map exactly to enum name. We could handle capitalization correction. Overall, I think either is okay. I do not have much of a preference.